Implement the LargestContentfulPaint interface

smfr opened the pull request:
#### 2eec14d73ded2bf390bb6ea1716937fa151e7a54
<pre>
Implement the LargestContentfulPaint interface
<a href="https://bugs.webkit.org/show_bug.cgi?id=297568">https://bugs.webkit.org/show_bug.cgi?id=297568</a>
<a href="https://rdar.apple.com/157604838">rdar://157604838</a>

Reviewed by Tim Nguyen and Ryosuke Niwa.

Add and implement the PaintTimingMixin and LargestContentfulPaint PerformanceEntry,
per <a href="https://w3c.github.io/largest-contentful-paint/#sec-largest-contentful-paint-interface">https://w3c.github.io/largest-contentful-paint/#sec-largest-contentful-paint-interface</a>,
adding the IDL files and implementing JS wrapper creation.

Exposure is gated on `Document::supportsLargestContentfulPaint()` which consults the Setting.

The performance entries are never created yet.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSPerformanceEntryCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::supportsLargestContentfulPaint const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LargestContentfulPaint.cpp: Added.
(WebCore::LargestContentfulPaint::LargestContentfulPaint):
(WebCore::LargestContentfulPaint::paintTime const):
(WebCore::LargestContentfulPaint::presentationTime const):
(WebCore::LargestContentfulPaint::loadTime const):
(WebCore::LargestContentfulPaint::renderTime const):
(WebCore::LargestContentfulPaint::size const):
(WebCore::LargestContentfulPaint::id const):
(WebCore::LargestContentfulPaint::url const):
(WebCore::LargestContentfulPaint::element const):
* Source/WebCore/page/LargestContentfulPaint.h: Copied from Source/WebCore/page/PerformanceEntry.cpp.
* Source/WebCore/page/LargestContentfulPaint.idl: Added.
* Source/WebCore/page/LocalDOMWindowProperty.cpp:
(WebCore::LocalDOMWindowProperty::frame const):
* Source/WebCore/page/PaintTimingMixin.idl: Added.
* Source/WebCore/page/PerformanceEntry.cpp:
(WebCore::PerformanceEntry::parseEntryTypeString):
* Source/WebCore/page/PerformanceEntry.h:
* Source/WebCore/page/PerformanceObserver.cpp:
(WebCore::PerformanceObserver::supportedEntryTypes):
* Source/WebCore/page/PointerCaptureController.h: Unified sources fix.

Canonical link: <a href="https://commits.webkit.org/299598@main">https://commits.webkit.org/299598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9fe67fc8a9812606de2f79e363f0b0b7d5dc183

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39182 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29837 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125761 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71565 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3d7be7c6-ba47-4751-a429-e912bcb6b2c7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121367 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47762 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90781 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/abaf680c-c037-4ff3-926a-a41fd88bb271) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122442 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31849 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107146 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71261 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/33677300-32d2-4a4c-b0b3-efe5e35bbfa4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30877 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25251 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69409 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101301 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25443 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128739 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46413 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35142 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99370 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103340 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99188 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25207 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44625 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22638 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42977 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46274 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51981 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45739 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49090 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47426 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/169184521d3aaacfeb51b96c26007b0758fb411f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f8d804414267213106c1cfadad5a92ba72002eb9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49565

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4346,6 +4346,11 @@ bool Document::supportsPaintTiming() const
     return protectedSecurityOrigin()->isSameOriginDomain(topOrigin());
 }
 
+bool Document::supportsLargestContentfulPaint() const
+{
+    return settings().largestContentfulPaintEnabled();
+}
```
I don't think this shortcut is really that much shorter, I would just inline `settings().largestContentfulPaintEnabled()` in the relevant places.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8d8a3aa8e5b35086b5f56969bac81b2a8d99e14d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f9fe67fc8a9812606de2f79e363f0b0b7d5dc183. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49565

smfr committed:
Implement the LargestContentfulPaint interface
https://bugs.webkit.org/show_bug.cgi?id=297568
rdar://157604838

Reviewed by Tim Nguyen and Ryosuke Niwa.

Add and implement the PaintTimingMixin and LargestContentfulPaint PerformanceEntry,
per https://w3c.github.io/largest-contentful-paint/#sec-largest-contentful-paint-interface,
adding the IDL files and implementing JS wrapper creation.

Exposure is gated on `Document::supportsLargestContentfulPaint()` which consults the Setting.

The performance entries are never created yet.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSPerformanceEntryCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::supportsLargestContentfulPaint const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LargestContentfulPaint.cpp: Added.
(WebCore::LargestContentfulPaint::LargestContentfulPaint):
(WebCore::LargestContentfulPaint::paintTime const):
(WebCore::LargestContentfulPaint::presentationTime const):
(WebCore::LargestContentfulPaint::loadTime const):
(WebCore::LargestContentfulPaint::renderTime const):
(WebCore::LargestContentfulPaint::size const):
(WebCore::LargestContentfulPaint::id const):
(WebCore::LargestContentfulPaint::url const):
(WebCore::LargestContentfulPaint::element const):
* Source/WebCore/page/LargestContentfulPaint.h: Copied from Source/WebCore/page/PerformanceEntry.cpp.
* Source/WebCore/page/LargestContentfulPaint.idl: Added.
* Source/WebCore/page/LocalDOMWindowProperty.cpp:
(WebCore::LocalDOMWindowProperty::frame const):
* Source/WebCore/page/PaintTimingMixin.idl: Added.
* Source/WebCore/page/PerformanceEntry.cpp:
(WebCore::PerformanceEntry::parseEntryTypeString):
* Source/WebCore/page/PerformanceEntry.h:
* Source/WebCore/page/PerformanceObserver.cpp:
(WebCore::PerformanceObserver::supportedEntryTypes):
* Source/WebCore/page/PointerCaptureController.h: Unified sources fix.

Canonical link: https://commits.webkit.org/299598@main
